scrnaseq.Rmd

```diff
@@ -1674,142 +1748,400 @@ if (degs_found) {
+  avg.2 = DegsAvgData(contrast[["object"]], cells=contrast[["cells_group2_idx_subset"]], genes=test_results$gene, slot=contrast[["slot"]])[,1]
+  test_results = cbind(test_results, avg.1, avg.2)
+  
+  # Add test results and drop unneccessary data
+  contrast = c(contrast, list(results=test_results))
+  contrast[["object"]] = NULL
+  contrast[["cells_group1_idx_subset"]] = NULL
+  contrast[["cells_group2_idx_subset"]] = NULL
+  
+  return(contrast)
+})
+
+# Also remove objects from deg_contrasts_list (save memory)
+degs_contrasts_list = purrr::map(degs_contrasts_list, function(contrast){ contrast[["object"]] = NULL; return(contrast)})
+
+# Compute enrichr results (not in parallel due to server load)
+degs_contrasts_results = purrr::map(degs_contrasts_results, function(contrast) {
+  # Get results table
+  results_table = contrast$results
+  
+  # Drop existing results
+  if ("enrichr" %in% names(contrast)) d[["enrichr"]] = NULL
+  
+  # Split into up- and downregulated DEGs, then translate to entrez gene, runEnrichr
+  degs_up = dplyr::filter(results_table, avg_log2FC>0) %>% dplyr::pull(gene) %>% unique()
+  degs_up = sapply(degs_up, function(n) seurat_rowname_to_entrez[[n]][1], USE.NAMES=TRUE, simplify=TRUE) %>% unlist() %>% as.character()
+  degs_up = degs_up[!is.na(degs_up)]
+  enrichr_results_up = EnrichrTest(genes=degs_up, databases=param$enrichr_dbs, padj=param$p_enrichr)
+  
+  degs_down = dplyr::filter(results_table, avg_log2FC<0) %>% dplyr::pull(gene) %>% unique()
+  degs_down = sapply(degs_down, function(n) seurat_rowname_to_entrez[[n]][1], USE.NAMES=TRUE, simplify=TRUE) %>% unlist() %>% as.character()
+  degs_down = degs_down[!is.na(degs_down)]
+  enrichr_results_down = EnrichrTest(genes=degs_down, databases=param$enrichr_dbs, padj=param$p_enrichr)
+
+  
+  # Flatten both enrichr results into tables
+  enrichr_results_up = purrr::map_dfr(names(enrichr_results_up), function(n) {
+    return(cbind(enrichr_results_up[[n]], 
+          list(Database=factor(rep(n, nrow(enrichr_results_up[[n]])), levels=names(enrichr_results_up)), 
+               Direction=factor(rep("up", nrow(enrichr_results_up[[n]])), levels=c("up", "down"))
+               )
+          ))
+  })
+  
+  enrichr_results_down = purrr::map_dfr(names(enrichr_results_down), function(n) {
+    return(cbind(enrichr_results_down[[n]], 
+          list(Database=factor(rep(n, nrow(enrichr_results_down[[n]])), levels=names(enrichr_results_down)), 
+               Direction=factor(rep("up", nrow(enrichr_results_down[[n]])), levels=c("up", "down"))
+               )
+          ))
+  })
+  
+  # Rbind and add factor levels
+  enrichr_results = dplyr::bind_rows(enrichr_results_up, enrichr_results_down)
+  return(c(contrast, list(enrichr=enrichr_results)))
+  
+  # Wait two second for server
+  Sys.sleep(2)
+})
+
+# Now regroup list so that subsets are together again
```

Comment:
Example: 

In param$deg_contrasts we can basically say: compare A vs B for clusters 1,2,3,4 and 5. That means that we subset all cells for cluster 1 and then compare, next subset all cells for cluster then compare, next .... .

In param$deg_contrasts, this is one row: condition   A   B    1,2,3,4,5. However for computation, I make five rows.

condition   A   B    1
condition   A   B    2
condition   A   B    3
condition   A   B    4
condition   A   B    5

After the computations are done, I simply regroup them again so that they organised together since they somehow belong together.